### ISSUE TITLE
0.8.0 setVariables change difficult when variables not yet updated

### ISSUE THREAD

**Issue by finneganh** [facebook/relay#1048]
I'm running into trouble due to this change:

“RelayContainer.setVariables will no longer check if the variables are changed before re-running the variables. To prevent extra work, check the current variables before calling setVariables.”

My problem is that my code had previously been doing a `setVariables` call from `componentWillReceiveProps`, which seems to (at least in my case) be called as a result of `setVariables`.

This is leading to an infinite loop, but I can’t use the suggestion to check the current value of the variable because it isn’t updated until the variable fetch succeeds.

While I can be more defensive about not considering `setVariables` based on what props are changed, I worry that this is more fragile.

Can you restore the previous `setVariables` behavior, or make it easier to see what `variables` values are in-flight so I can make my own check more accurate?

**Comment by josephsavona** [comment]
Thanks for letting us know about this issue. In general, calling `setVariables` in `componentWillReceiveProps` can lead to an infinite loop. Changing variables fetches more data, which will typically cause the component to re-render, which will cause new props to be received. I'll get to a workaround/follow-up in a moment, but before that - what's the use-case for setting variables in `willReceiveProps`? If a component needs data, that *should* be expressible in its fragments such that the parent can ensure all data is available.

We encountered some issue with the old behavior of `setVariables` and don't plan on reverting the old behavior. Specifically, it's possible for one component in a to call `forceFetch` such that another component ends up missing some data (forceFetch on a range overwrites the range - if i force fetch for 5 items and you had fetched 20, now you'll suddenly get only 5. we refer to this as the "missing records/items problem"). The only option before the setVars change was for the affected component to also forceFetch (because setVars would previously see that the affected components vars were the same and avoid refetching), which could cause an infinite loop of refetching. The new behavior allows a component so affected to use `setVars` and avoid the infinite loop (because setVars merges new range items instead of overwriting).

tl;dr Yes, we should make it possible to see what vars are pending, but I'm still curious about the use-case for calling `setVariables` in willReceiveProps (seems like an anti-pattern).

cc @yuzhi

**Comment by josephsavona** [comment]
I've filed a new issue at #1049 to describe the addition of `pendingVariables`.

**Comment by wincent** [comment]
Also curious to know what about your use case requires you to use `componentWillReceiveProps` (as opposed to, say, `componentDidUpdate`).

**Comment by finneganh** [comment]
Thanks for the info about the problems with the previous implementation!  Here's what we're up to, and maybe you'll have a suggestion for a better approach.


We're rendering a list of elements with a filtering search box. We're doing the filtering client-side, so we want to load the complete list into memory (up to at least a big max of 5000 elements or so). For keeping the individual requests reasonable, though (we have resource limits on the server to mitigate abuse), we're making requests for 100 or so at a time.

So, the logic is that we have a "count" variable that feeds into the "first" argument of the elements' connection. When props update, we check the connection to see if it has a next page. If so, we set the variable to 100 + the current number of edges to fetch the next page.


We'd probably be happier either doing this in parallel, or always setting "first" to the 5000 max, with some combination of server-side limit enforcement and Relay automatically making follow-up requests to get more pages.

It might also be that `componentDidUpdate` would work? We don't actually render all of the loaded elements by default; they're just in memory for the filtering.

**Comment by josephsavona** [comment]
@finneganh I see, thanks for clarifying. In general we recommend doing this type of fetching in `componendDidMount`. You might set things up as:


```javascript
componentDidMount() {
  // begin loading more only once the component has mounted
  this.fetchMore();
}
fetchMore() {
  if (!this.props.results.pageInfo.hasNextPage) {
    return; // nothing more to fetch
  }
  const {count} = this.props.relay.variables;
  this.props.relay.setVariables({count: count: 100}, {done} => {
    if (done) {
      // continue loading when the previous fetch completes
      this.fetchMore();
    }
  });
}
```

**Comment by finneganh** [comment]
Ah, that looks tons cleaner. Thanks for the tip, I'll use that approach.

Closing this ticket because you're tracking `pendingVariables` elsewhere.

**Comment by josephsavona** [comment]
@finneganh Glad that helps, thanks for the follow-up!

**Comment by jardakotesovec** [comment]
I am having similar problem. My use case is if I get new props from parent I need to adjust relay variables accordingly . So I thought that `componentWillRecieveProps` or `componentDidUpdate` would be good place to do that. 

But issue is with the infinite loop  (with both 0.7.3 and 0.8) - seems that component gets updated immediately after calling setVariables, which causes next setVariable call..  and it does not wait for the actual requests, just generate as much requests as it can in infinite loop.

I am not able to reproduce it in relay playground so I am still in process figuring out whats going on. Just in case someone has insight and/or dealing with this as well...

**Comment by jardakotesovec** [comment]
So here is [playground](https://facebook.github.io/relay/prototyping/playground.html#source=%2F**%0A%20*%20Copyright%20(c)%202013-present%2C%20Facebook%2C%20Inc.%0A%20*%20All%20rights%20reserved.%0A%20*%0A%20*%20This%20source%20code%20is%20licensed%20under%20the%20BSD-style%20license%20found%20in%20the%0A%20*%20LICENSE%20file%20in%20the%20root%20directory%20of%20this%20source%20tree.%20An%20additional%20grant%0A%20*%20of%20patent%20rights%20can%20be%20found%20in%20the%20PATENTS%20file%20in%20the%20same%20directory.%0A%20*%2F%0A%0A%2F%2F%20Welcome%20to%20Relay.%0A%2F%2F%20Allow%20us%20to%20introduce%20you%20to%20the%20four%20elements.%0A%0A%2F**%0A%20*%20%231%20-%20Your%20React%20components%0A%20*%20This%20will%20look%20familiar%20to%20React%20developers.%0A%20*%0A%20*%20To%20learn%20more%20about%20React%2C%20visit%3A%0A%20*%20%20https%3A%2F%2Ffacebook.github.io%2Freact%0A%20*%2F%0Aclass%20HelloApp%20extends%20React.Component%20%7B%0A%20%20componentDidMount()%20%7B%0A%20%20%20%20console.log('componentDidMount')%0A%20%20%20%20this.props.relay.setVariables(%7B%0A%20%20%20%20%20%20bar%3A%202%0A%20%20%20%20%7D)%0A%0A%20%20%7D%0A%20%20componentWillReceiveProps()%20%7B%0A%20%20%20%20console.log('componentWillReceiveProps')%3B%0A%0A%20%20%7D%0A%20%20componentDidUpdate()%20%7B%0A%20%20%20%20console.log('componentDidUpdate')%3B%0A%20%20%20%20this.props.relay.setVariables(%7B%0A%20%20%20%20%20%20bar%3A%202%0A%20%20%20%20%7D)%0A%0A%20%20%7D%20%0A%20%20render()%20%7B%0A%20%20%20%20%2F%2F%20Relay%20will%20materialize%20this%20prop%20based%20on%20the%0A%20%20%20%20%2F%2F%20result%20of%20the%20query%20in%20the%20next%20component.%0A%20%20%20%20const%20%7Bhello%7D%20%3D%20this.props.greetings%3B%0A%20%20%20%20return%20%3Ch1%3E%7Bhello%7D%3C%2Fh1%3E%3B%0A%20%20%7D%0A%7D%0A%20%0A%2F**%0A%20*%20%232%20-%20Relay%20containers%0A%20*%20Compose%20your%20React%20components%20with%20a%20declaration%20of%0A%20*%20the%20GraphQL%20query%20fragments%20that%20fetch%20their%20data.%0A%20*%0A%20*%20To%20learn%20more%20about%20Relay%20containers%2C%20visit%3A%0A%20*%20%20%20https%3A%2F%2Ffacebook.github.io%2Frelay%2Fdocs%2Fguides-containers.html%0A%20*%2F%0AHelloApp%20%3D%20Relay.createContainer(HelloApp%2C%20%7B%0A%20%20initialVariables%3A%20%7B%0A%20%20%20%20bar%3A%201%0A%20%20%7D%2C%0A%20%20fragments%3A%20%7B%0A%20%20%20%20%2F%2F%20This%20GraphQL%20query%20executes%20against%0A%20%20%20%20%2F%2F%20the%20schema%20in%20the%20'schema'%20tab%20above.%0A%20%20%20%20%2F%2F%0A%20%20%20%20%2F%2F%20To%20learn%20more%20about%20Relay.QL%2C%20visit%3A%0A%20%20%20%20%2F%2F%20%20%20https%3A%2F%2Ffacebook.github.io%2Frelay%2Fdocs%2Fapi-reference-relay-ql.html%0A%20%20%20%20greetings%3A%20()%20%3D%3E%20Relay.QL%60%0A%20%20%20%20%20%20fragment%20on%20Greetings%20%7B%0A%20%20%20%20%20%20%20%20hello%20(bar%3A%20%24bar)%0A%20%20%20%20%20%20%7D%20%0A%20%20%20%20%60%2C%0A%20%20%7D%0A%7D)%3B%0A%0A%2F**%0A%20*%20%233%20-%20Relay%20routes%0A%20*%20Define%20a%20root%20GraphQL%20query%20into%20which%20your%0A%20*%20containers'%20query%20fragments%20will%20be%20composed.%0A%20*%0A%20*%20To%20learn%20more%20about%20Relay%20routes%2C%20visit%3A%0A%20*%20%20%20https%3A%2F%2Ffacebook.github.io%2Frelay%2Fdocs%2Fguides-routes.html%0A%20*%2F%0Aclass%20HelloRoute%20extends%20Relay.Route%20%7B%0A%20%20static%20routeName%20%3D%20'Hello'%3B%20%20%2F%2F%20A%20unique%20name%0A%20%20static%20queries%20%3D%20%7B%0A%20%20%20%20%2F%2F%20Here%2C%20we%20compose%20your%20Relay%20container's%0A%20%20%20%20%2F%2F%20'greetings'%20fragment%20into%20the%20'greetings'%0A%20%20%20%20%2F%2F%20field%20at%20the%20root%20of%20the%20GraphQL%20schema.%0A%20%20%20%20greetings%3A%20(Component)%20%3D%3E%20Relay.QL%60%0A%20%20%20%20%20%20query%20GreetingsQuery%20%7B%0A%20%20%20%20%20%20%20%20greetings%20%7B%0A%20%20%20%20%20%20%20%20%20%20%24%7BComponent.getFragment('greetings')%7D%2C%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%60%2C%0A%20%20%7D%3B%0A%7D%0A%0A%2F**%0A%20*%20%234%20-%20Relay%20root%20containers%0A%20*%20Compose%20a%20Relay%20container%20with%20a%20Relay%20route.%0A%20*%20This%20enables%20Relay%20to%20synthesize%20a%20complete%20query%0A%20*%20to%20fetch%20the%20data%20necessary%20to%20render%20your%20app.%0A%20*%0A%20*%20To%20learn%20more%20about%20Relay%20root%20containers%2C%20visit%3A%0A%20*%20%20%20https%3A%2F%2Ffacebook.github.io%2Frelay%2Fdocs%2Fguides-root-container.html%0A%20*%2F%0AReactDOM.render(%0A%20%20%3CRelay.RootContainer%0A%20%20%20%20Component%3D%7BHelloApp%7D%0A%20%20%20%20route%3D%7Bnew%20HelloRoute()%7D%0A%20%20%20%20renderFetched%3D%7Bfunction(data)%20%7B%0A%20%20%20%20return%20(%0A%20%20%20%20%20%20%3CHelloApp%20%7B...data%7D%20breakMe%3D%7B5%7D%2F%3E%0A%20%20%20%20)%3B%0A%20%20%7D%7D%0A%0A%20%20%2F%3E%2C%0A%20%20mountNode%0A)%3B%0A&schema=%2F**%0A%20*%20Copyright%20(c)%202013-present%2C%20Facebook%2C%20Inc.%0A%20*%20All%20rights%20reserved.%0A%20*%0A%20*%20This%20source%20code%20is%20licensed%20under%20the%20BSD-style%20license%20found%20in%20the%0A%20*%20LICENSE%20file%20in%20the%20root%20directory%20of%20this%20source%20tree.%20An%20additional%20grant%0A%20*%20of%20patent%20rights%20can%20be%20found%20in%20the%20PATENTS%20file%20in%20the%20same%20directory.%0A%20*%2F%0A%0Aimport%20%7B%0A%20%20GraphQLObjectType%2C%0A%20%20GraphQLSchema%2C%0A%20%20GraphQLString%2C%0A%20%20%20%20GraphQLInt%0A%0A%7D%20from%20'graphql'%3B%0A%0Aconst%20GREETINGS%20%3D%20%7B%0A%20%20hello%3A%20'Hello%20world'%2C%0A%7D%3B%0A%0A%2F**%0A%20*%20Objects.%0A%20*%20Build%20up%20a%20portrait%20of%20your%20data%20universe%0A%20*%20using%20the%20object%20type.%20Here%2C%20we%20define%20a%0A%20*%20type%20of%20object%20that%20has%20a%20'hello'%20field%0A%20*%20that%20is%20of%20the%20string%20type.%0A%20*%2F%0Aconst%20GreetingsType%20%3D%20new%20GraphQLObjectType(%7B%0A%20%20name%3A%20'Greetings'%2C%0A%20%20fields%3A%20()%20%3D%3E%20(%7B%0A%20%20%20%20hello%3A%20%7B%0A%20%20%20%20%20%20type%3A%20GraphQLString%2C%0A%20%20%20%20%20%20args%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20bar%3A%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20type%3A%20GraphQLInt%0A%20%20%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%7D%2C%0A%20%20%7D)%2C%0A%7D)%3B%0A%0A%2F**%0A%20*%20The%20schema.%0A%20*%20Here%20we%20export%20a%20schema%20that%20offers%20one%20root%0A%20*%20field%20named%20'greetings'%2C%20and%20a%20method%20to%0A%20*%20resolve%20its%20data.%0A%20*%0A%20*%20To%20learn%20more%20about%20writing%20GraphQL%20schemas%20for%20Relay%2C%20visit%3A%0A%20*%20%20%20https%3A%2F%2Fgithub.com%2Fgraphql%2Fgraphql-relay-js%0A%20*%2F%0Aexport%20default%20new%20GraphQLSchema(%7B%0A%20%20query%3A%20new%20GraphQLObjectType(%7B%0A%20%20%20%20name%3A%20'Query'%2C%0A%20%20%20%20fields%3A%20()%20%3D%3E%20(%7B%0A%20%20%20%20%20%20greetings%3A%20%7B%0A%20%20%20%20%20%20%20%20type%3A%20GreetingsType%2C%0A%20%20%20%20%20%20%20%20%2F%2F%20Here%20we%20define%20a%20resolver%20that%20returns%0A%20%20%20%20%20%20%20%20%2F%2F%20the%20data%20defined%20above.%20Were%20this%20schema%0A%20%20%20%20%20%20%20%20%2F%2F%20executing%20on%20the%20server%20side%2C%20you%20could%0A%20%20%20%20%20%20%20%20%2F%2F%20write%20a%20resolve%20method%20that%20fetches%0A%20%20%20%20%20%20%20%20%2F%2F%20live%20data%20from%20a%20database.%0A%20%20%20%20%20%20%20%20resolve%3A%20()%20%3D%3E%20GREETINGS%2C%0A%20%20%20%20%20%20%7D%2C%0A%20%20%20%20%7D)%2C%0A%20%20%7D)%2C%0A%7D)%3B%0A).

To achieve infinite loop in this playground - only thing that needed is to change `breakMe={5}` to something non-scalar like  `breakMe={{}}`. As result `shouldComponentUpdate` in RelayContainer always returns true which is the issue because `setState` get called on container immediately after setVariable.

So I guess solution here is to implement proper `shouldComponentUpdate` in component itself and call setVariables in `componentDidUpdate`? And in 0.8 I should also make the comparison with current variables (which will be possible once we have [pending variable prop](https://github.com/facebook/relay/issues/1049)) before I call new setVariables?

Currently feel overwhelmed to do relatively simple thing. Am I doing it wrong?

**Comment by josephsavona** [comment]
@jardakotesovec Thanks for providing more detail here. This came up again in #1063 (see my [comment](https://github.com/facebook/relay/issues/1063#issuecomment-213025165)). It sounds like we need to adjust RelayContainer such that it's safe to call `setVariables` from within the child's `componentWillReceiveProps` call. 

We'll look at this as time permits, but are definitely open to PRs to help here.

**Comment by mjhlybmwq** [comment]
@jardakotesovec I think you can check your `props` before `setVariables` is called in `componentWillReceiveProps` to break the loop. For example:
```javascript
componentWillReceiveProps(nextProps){
    if (this.props.foo !== nextProps.foo){
        this.props.relay.setVariables({...})
    } 
}
```

**Comment by jardakotesovec** [comment]
@mjhlybmwq yes.. or implement `shouldComponentUpdate` and call `setVariables` in `componentDidUpdate`. But sticking with 0.7.3 for now..  until I can check pending variables if I should `setVariables`

**Comment by finneganh** [comment]
@josephsavona Just wanted to drop in our solution here to show what we ended up going with, in case it might be informative for future Relay features: https://gist.github.com/finneganh/c81bd1167a07306eba7365f8e123ccfb

One note about the suggestion that you put above is that at the time of the `setVariables` callback’s ready state being “done,” the component’s props haven’t been updated and so `fetchMore` wouldn’t see the right view of the world.

We ended up putting things through `componentDidUpdate`, with some checks to both prevent double-fetching and also a backoff if the fetches are failing or otherwise not making progress (so as not to fetch on an infinite loop, especially during prod issues).

One curiosity we had was that for our prod component, `componentWillReceiveProps` was consistently called (twice actually) after the `setVariables` call errored out. In a smaller test component, it wasn’t called. I don’t know if there’s an explanation for that behavior (the prod component had sub-containers, maybe that was it?) but we ended up working around it with the `forceUpdate`.

At any rate, I found this an interesting case where we seemed to be running at odds with how Relay wants to do data access. I don’t know if that means we’re doing something particularly wrong, or just running into the law of leaky abstractions.
